rust/core-lib/src/selection.rs

```diff
@@ -109,16 +109,63 @@ impl Selection {
         &self.regions[first..last]
     }
 
-    /// Deletes all the regions that intersect or touch the given range.
-    pub fn delete_range(&mut self, start: usize, end: usize) {
-        let first = self.search(start);
+    /// Deletes all the regions that intersect or (if delete_adjacent = true) touch the given range.
+    pub fn delete_range(&mut self, start: usize, end: usize, delete_adjacent: bool) {
+        let mut first = self.search(start);
         let mut last = self.search(end);
-        if last < self.regions.len() && self.regions[last].min() <= end {
+        if first >= self.regions.len() {
+            return;
+        }
+        if !delete_adjacent && self.regions[first].max() == start {
+            first += 1;
+        }
+        if last < self.regions.len() && ((delete_adjacent && self.regions[last].min() <= end)
+           || (!delete_adjacent && self.regions[last].min() < end)) {
             last += 1;
         }
         remove_n_at(&mut self.regions, first, last - first);
     }
 
+    /// Add a region to the selection. This method does not merge regions and does not allow
+    /// ambiguous regions (regions that overlap).
+    ///
+    /// On ambiguous regions, the regions closer to the left (to the file beginning) wins. That is,
```

Comment:
language nit: although this is currently used primarily to deal with ranges of lines in a file, the idea of a 'file' is external to this data structure. I'm also unsure about the word 'left'. Where possible I would prefer something like 'lower', e.g. something that references a property of the value, not a property of a particular conception of the number line.